rascaline/src/calculators/lode/spherical_expansion.rs

```diff
@@ -92,48 +78,63 @@ pub struct LodeSphericalExpansion {
+    real: Array2<f64>,
+    /// Imaginary part of `e^{i k r}`, the array shape is `(n_atoms, k_vector)`
+    imag: Array2<f64>,
+    /// Real part of `\sum_j e^{i k r_i} e^{-i k r_j}`, with one map entry for
+    /// each species of the atom j. The arrays shape are `(n_atoms, k_vector)`
+    real_per_center: BTreeMap<i32, Array2<f64>>,
+    /// Imaginary part of `\sum_j e^{i k r_i} e^{-i k r_j}`, with one map entry
+    /// for each species of the atom j. The arrays shape are `(n_atoms,
+    /// k_vector)`
+    imag_per_center: BTreeMap<i32, Array2<f64>>,
 }
 
-fn compute_structure_factors(positions: &[Vector3D], k_vectors: &[KVector]) -> StructureFactors {
-    let num_atoms: usize = positions.len();
-    let num_kvecs: usize = k_vectors.len();
+fn compute_structure_factors(positions: &[Vector3D], species: &[i32], k_vectors: &[KVector]) -> StructureFactors {
+    let n_atoms = positions.len();
+    let n_k_vectors = k_vectors.len();
 
-    let mut cosines = Array2::from_elem((num_kvecs, num_atoms), 0.0);
-    let mut sines = Array2::from_elem((num_kvecs, num_atoms), 0.0);
-
-    // cosines[i, j] = cos(k_i * r_j), same for sines
-    for (ik, k_vector) in k_vectors.iter().enumerate() {
-        for (atom_i, position) in positions.iter().enumerate() {
-            // dot product between kvectors and positions
+    let mut cosines = Array2::from_elem((n_atoms, n_k_vectors), 0.0);
+    let mut sines = Array2::from_elem((n_atoms, n_k_vectors), 0.0);
+    for (atom_i, position) in positions.iter().enumerate() {
+        for (ik, k_vector) in k_vectors.iter().enumerate() {
+            // dot product between k-vector and positions
             let s = k_vector.norm * k_vector.direction * position;
 
-            cosines[[ik, atom_i]] = f64::cos(s);
-            sines[[ik, atom_i]] = f64::sin(s);
+            cosines[[atom_i, ik]] = f64::cos(s);
+            sines[[atom_i, ik]] = f64::sin(s);
         }
     }
 
-    let mut structure_factors = StructureFactors {
-        real: Array3::from_elem((num_atoms, num_atoms, num_kvecs), 0.0),
-        imag: Array3::from_elem((num_atoms, num_atoms, num_kvecs), 0.0),
-    };
-
-    for i in 0..num_atoms {
-        for j in 0..num_atoms {
-            for k in 0..num_kvecs {
-                structure_factors.real[[i, j, k]] = cosines[[k, i]] * cosines[[k, j]] + sines[[k, i]] * sines[[k, j]];
-                structure_factors.imag[[i, j, k]] = sines[[k, i]] * cosines[[k, j]] - cosines[[k, i]] * sines[[k, j]];
+    let all_species = species.iter().copied().collect::<BTreeSet<_>>();
+    let mut real_per_center = all_species.iter().copied()
+        .map(|s| (s, Array2::from_elem((n_atoms, n_k_vectors), 0.0)))
+        .collect::<BTreeMap<_, _>>();
+    let mut imag_per_center = all_species.iter().copied()
+        .map(|s| (s, Array2::from_elem((n_atoms, n_k_vectors), 0.0)))
+        .collect::<BTreeMap<_, _>>();
+
+    for i in 0..n_atoms {
+        for j in 0..n_atoms {
+            for k in 0..n_k_vectors {
```

Comment:
For parallelism the plan is to use also the `.iter()` here as well instead of the explicit loop over `k`, right?

Reply:
We can try making the structure factor calculation parallel as well yes! If you want to give it a go feel free to